include/sys/queued_operation.h

```diff
@@ -0,0 +1,343 @@
+	 * process() be re-entrant.  Generally the process() function
+	 * should itself be async, with finalization done after
+	 * process() returns.
+	 *
+	 * @note The process function must be isr-ok.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation that should be initiated.  A null
+	 * pointer is passed if there are no pending operations.
+	 */
+	void (*process)(struct queued_operation_manager *mgr,
+			struct queued_operation *op);
+};
+
+/**
+ * @brief State associated with a manager instance.
+ */
+struct queued_operation_manager {
+	/* Links the operation into the operation queue. */
+	sys_slist_t operations;
+
+	/* Pointer to the functions that support the manager. */
+	const struct queued_operation_functions *vtable;
+
+	/* Pointer to an on-off service supporting this service.  NULL
+	 * if service is always available.
+	 */
+	struct onoff_manager *onoff;
+
+	/* The state of on-off service requests. */
+	struct onoff_client onoff_client;
+
+	/* Lock controlling access to other fields. */
+	struct k_spinlock lock;
+
+	/* The operation that is being processed. */
+	struct queued_operation *current;
+
+	/* Information about the internal state of the manager. */
+	uint32_t volatile state;
+};
+
+#define QUEUED_OPERATION_MANAGER_INITIALIZER(_vtable, _onoff) {	\
+		.vtable = _vtable,				\
+		.onoff = _onoff,				\
+}
+
+/**
+ * @brief Submit an operation to be processed when the service is
+ * available.
+ *
+ * The service process function will be invoked during this call if
+ * the service is available.
+ *
+ * @param mgr a generic pointer to the service instance
+ *
+ * @param op a generic pointer to an operation to be performed.  The
+ * notify field in the provided operation must have been initialized
+ * before being submitted, even if the operation description is being
+ * re-used.  This may be done directly with sys_notify API or by
+ * wrapping it in a service-specific operation init function.
+ *
+ * @param priority the priority of the operation relative to other
+ * operations.  Numerically lower values are higher priority.  Values
+ * outside the range of a signed 8-bit integer will be rejected,
+ * except for named priorities like QUEUED_OPERATION_PRIORITY_APPEND.
+ *
+ * @retval -ENOTSUP if callback notification is requested and the
+ * service does not provide a callback translation.  This may also be
+ * returned due to service-specific validation.
+ *
+ * @retval -EINVAL if the passed priority is out of the range of
+ * supported priorities.  This may also be returned due to
+ * service-specific validation.
+ *
+ * @return A negative value if the operation was rejected by service
+ * validation or due to other configuration errors.  A non-negative
+ * value indicates the operation has been accepted for processing and
+ * completion notification will be provided.
+ */
+int queued_operation_submit(struct queued_operation_manager *mgr,
```

Comment:
i was wondering about possibility to detect that given operation is already enqueued and return error (`-EBUSY`?). Maybe node field could be used for that (set to null when op is idle).
I'm having following scenario: sensor which has one queued operation instance which is used to schedule reading and i won't to detect situation when someone schedules next read (reusing same queued operation) before previous one is completed.

Reply:
That changes the contract used consistently for the resource management APIs: the client (in this case the sensor acting on behalf of a caller) submits a request, which is then owned by the manager until something notifies the client that the manager has released the request.

For your use case the sensor must be aware that there's an in-progress operation and refuse the attempt to schedule another.